lib/mongoose_push_web/schemas/request/send_notification/flat_notification.ex

```diff
@@ -45,4 +46,45 @@ defmodule MongoosePushWeb.Schemas.Request.SendNotification.Flat do
     },
     additionalProperties: false
   })
+
+  defimpl MongoosePushWeb.Protocols.RequestDecoder,
+    for: Request.SendNotification.FlatNotification do
+    @spec decode(%Request.SendNotification.FlatNotification{}) :: MongoosePush.request()
+    def decode(schema) do
+      %{
+        service: String.to_atom(schema.service),
+        alert: %{
+          body: schema.body,
+          title: schema.title
+        }
+      }
+      |> add_optional_fields(schema)
+      |> add_optional_alert_fields(schema)
+    end
+
+    defp add_optional_fields(push_request, schema) do
+      opt_keys = [:data, :mode, :topic]
+
+      Enum.reduce(opt_keys, push_request, fn x, acc ->
+        case Map.get(schema, x) do
+          nil -> acc
+          val -> Map.put(acc, x, maybe_convert_to_atom(x, val))
+        end
+      end)
+    end
+
+    defp add_optional_alert_fields(push_request, schema) do
+      opt_alert_keys = [:badge, :click_action, :tag]
+
+      Enum.reduce(opt_alert_keys, push_request, fn x, acc ->
+        case Map.get(schema, x) do
+          nil -> acc
+          val -> Kernel.update_in(acc, [:alert, x], fn _ -> val end)
+        end
+      end)
+    end
+
+    defp maybe_convert_to_atom(:mode, val), do: String.to_atom(val)
```

Comment:
As far as I know, we are converting here any input from the user to an atom. I consider this unsafe. From what I remember we accept only two values (`prod` and `dev`) here so maybe we could just parse the input and return atom for these values we accept and rise and error or do sth esle appropriate for invalid values.

Reply:
`String.to_existing_atom/1`

Reply:
Even more sth like:

Reply:
thank you @michalwski for this catch